MAINT: add support for a==0 and x>0 edge case to igam and igamc

rlucas7 opened the pull request:
#### Reference issue
Closes #10948 

#### What does this implement/fix?
Adds the edge case of `a==0 && x>0` to not emit  `nan` in `igamc` and `igam` and adds unit tests for the new values.

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -129,9 +129,15 @@ double igam(double a, double x)
+    if (x < 0 || a < 0) {
 	sf_error("gammainc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
```
Let's also move up the `if (x == 0)` branch from below to keep all the various-parameters-zero logic clustered together.

rlucas7 replied:
done

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -129,9 +129,15 @@ double igam(double a, double x)
 	sf_error("gammainc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
+	    if(x > 0) {
```
Space between the `if` and the `(`.

rlucas7 replied:
done, now reads `if (x > 0) {`

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
 	sf_error("gammaincc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
+        if ( x > 0 ) {
```
Sub `(x > 0)` for `( x > 0 )`.

rlucas7 replied:
done

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
+    if (x < 0 || a < 0) {
 	sf_error("gammaincc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
```
As above about moving the `x == 0` branch up.

rlucas7 replied:
done

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
+        if ( x > 0 ) {
+		return 0;
+	} else {
+		return NPY_NAN;
```
Looks like your spacing settings are off-we want to indent by 4 spaces.

rlucas7 replied:
not sure how the tabs got in there, fixed

rlucas7 commented:
> Thanks for picking this one up @rlucas7. I would recommend also adding the following tests:
> 
>     * Check that you get `nan` at `a = 0, x = 0`.
> 
>     * Add a "limit check test", e.g. that `sc.gammaincc(1e-10, 1)` converges to the special case where `a = 0` and `x = 1`.

I've added tests for both `sc.gammainc()` and `sc.gammaincc()` for limit check and for `nan` case so should be good on that end. Also, updated the style/formatting changes. Thanks for taking a look @person142  let's see if these changes pass on CI, they passed locally.

person142 commented on scipy/special/tests/test_gammainc.py:
```diff
@@ -53,6 +55,13 @@ def test_x_zero(self):
         assert_array_equal(sc.gammainc(a, 0), 0)
 
+    def test_nan_at_a_and_x_0(self):
+        assert np.isnan(sc.gammainc(0,0))
```
You should be able to just add `(0, 0)` to the `INVALID_POINTS` list above.

rlucas7 replied:
done

person142 commented on scipy/special/tests/test_gammainc.py:
```diff
@@ -53,6 +55,13 @@ def test_x_zero(self):
+        assert np.isnan(sc.gammainc(0,0))
+
+    def test_limit_check(self):
+        result = sc.gammainc(1e-10, 1)
```
I would assert that `sc.gammainc(1e-10, 1)` is close to `sc.gammainc(0, 1)`.

rlucas7 replied:
done.

person142 commented on scipy/special/tests/test_gammainc.py:
```diff
@@ -111,6 +123,14 @@ def test_infinite_limits(self):
+
+    def test_limit_check(self):
+        limit = sc.gammaincc(1e-10,1)
+        assert np.isclose(limit, 0)
```
Same comment as above.

rlucas7 replied:
done

person142 commented on scipy/special/tests/test_gammainc.py:
```diff
@@ -111,6 +123,14 @@ def test_infinite_limits(self):
             rtol=0
         )
 
+    def test_nan_a_and_x_0(self):
```
Same comment as above.

rlucas7 replied:
same response as above, no need to test twice :)

person142 commented on scipy/special/cephes/igam.c:
```diff
@@ -129,19 +129,25 @@ double igam(double a, double x)
-	    return NPY_NAN;
-	}
-	return 0;
+    if(x < 0 || a < 0) {
```
Space after the `if`.

rlucas7 replied:
done

rlucas7 commented:
@person142 the 2 failing builds are unrelated, the travis builds fail because of failing tests in linear program routines which haven't been touched by this PR.

person142 commented:
> the 2 failing builds are unrelated

Reran them and they passed on the second try.
